hooks/image/70-flatpak-appstream-catalog

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+
+# Construct AppStream XML for preinstalled apps
+
+import glob
+import logging
+import os
+import subprocess
+
+logging.basicConfig(
+    level=logging.INFO,
+    format='+ %(asctime)s %(levelname)s %(name)s: %(message)s',
+    datefmt='%H:%M:%S'
+)
+logger = logging.getLogger(os.path.basename(__file__))
+
+
+def main():
+    root = os.path.join(os.environ['OSTREE_VAR'], 'lib/flatpak/exports')
+
+    # Find the metainfo files to include in the AppStream.
+    ids = []
+    for metainfo in glob.glob(os.path.join(root, 'share/metainfo/*.xml')):
+        # Everything here should end in .metainfo.xml or .appdata.xml, but
+        # let's be sure.
+        if not (metainfo.endswith('.metainfo.xml') or
+                metainfo.endswith('.appdata.xml')):
+            raise Exception(f'Unexpected metainfo file {metainfo}')
+
+        # It's kind of bad to reject these files here because they're obviously
+        # valid enough for Flathub to have published them. Reasons that appstream-util
+        # 0.7.18-1+deb11u1 considers appdata to be invalid include:
+        #
+        # - Using a <url type=''> attribute from a more recent version of the spec
+        #   (dozens of apps, including GIMP)
+        #
+        # - A <li> element contains a nested <em> element (Pods; OK this is not a
+        #   preinstalled app, but still)
+        #
+        # - <release> elements are not in order (FileRoller)
+        #
+        # None of these are reasons to exclude the app from the catalogue but there is
+        # no way to tell the tool to ignore them: if we pass these files to
+        # appstream-compose it will fail.
+        #
+        # TODO: Switch to appstreamcli validate-relax once we are on Bookworm and can
+        # use appstreamcli compose.
```

Comment:
The other way to do this is not to use compose at all but to just build an XML catalog manually like [here](https://github.com/endlessm/gnome-software-data/blob/master/app-info/generate-eos-extra-appstream). I don't know if you'll run into the same post-catalog validation problems, though. Although, in that case you can just use `appstreamcli validate-relax`. I'm pretty sure flatpak [builds the appstream catalog manually](https://github.com/flatpak/flatpak/blob/14d735bfdaf211d13b7f0b4b9f203b847a581f39/common/flatpak-utils.c#L5354) internally.

Reply:
I think you're right. This also avoids stuff like dealing with icons.

Reply:
It turns out not to be an option to build the XML catalog from files in `/var/lib/flatpak/exports/share/metainfo` because, despite using the newer directory name `metainfo`, files in this directory (I'm looking at you, Firefox) may be in the old pre-AppStream 0.6 format where the top-level element is not a `<component>`. appstream-glib knows how to read these and migrate them to the new format.

The next stop was to try iterating through `/var/lib/flatpak/app/...../files/share/app-info/xmls` which are the individual `<component>`s that get concatenated by `flatpak build-update-repo` to make the repo appstream. But apart from having to duplicate Flatpak's logic to only export the one(s? not sure) matching the app ID, this also means you don't get the `<bundle type="flatpak">app/ID/ARCH/BRANCH</bundle>` tag, which is actually the only reliable way to map from a Flatpak ID to an entry in an AppStream catalog. These `<bundle>` tags are only added during `flatpak build-update-repo`.

And then inspiration struck: I have been approaching this from the wrong direction. `/var/lib/flatpak/appstream/$REMOTE/$ARCH/active/appstream.xml` contains a superset of the information we want; so we just have to filter those by the `<bundle>` tag.

Reply:
Right. I'd been hoping to avoid digging through all the remote appstream catalogs, but I guess if that's the only way...

Reply:
Actually I think this is optimal because at each stage in the build process more information is added to the data for each app; so by going to the remote catalogues we are getting the most-refined form of the data!